### PR TITLE
A basic version of SGEN Tool

### LINKED ISSUE #17580: Support SGEN in core
Add SGEN function on the project @17579

### PR DESCRIPTION
**Pull request by huanwu**
Fix #17580 

@shmao @mconnew @zhenlan

### CODE REVIEW

**Comment by huanwu** [comment]
@dotnet-bot test Innerloop OSX10.12 Debug x64 Build and Test

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs [hunk 1]
```diff
@@ -238,6 +239,182 @@ internal static Assembly LoadGeneratedAssembly(Type type, string defaultNamespac
...
+                compiler.AddImport(typeof(object).Assembly);
+                compiler.AddImport(typeof(XmlSerializer).Assembly);
+
+                IndentedWriter writer = new IndentedWriter(compiler.Source, false);
+
+                writer.WriteLine("#if _DYNAMIC_XMLSERIALIZER_COMPILATION");
+                writer.WriteLine("[assembly:System.Security.AllowPartiallyTrustedCallers()]");
+                writer.WriteLine("[assembly:System.Security.SecurityTransparent()]");
+                writer.WriteLine("[assembly:System.Security.SecurityRules(System.Security.SecurityRuleSet.Level1)]");
+                writer.WriteLine("#endif");
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: Line 305 ~ line 309 are not necessary on NetCore as security attributes are no-ops on NetCore. We can remove them.

File: src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs [hunk 2]
```diff
@@ -788,6 +788,46 @@ public static XmlSerializer[] FromMappings(XmlMapping[] mappings, Type type)
...
+            Assembly assembly = null;
+            for (int i = 0; i < types.Length; i++)
+            {
+                Type type = types[i];
+                if (DynamicAssemblies.IsTypeDynamic(type))
+                {
+                    throw new InvalidOperationException(SR.Format(SR.XmlPregenTypeDynamic, type.FullName));
+                }
+                if (assembly == null)
+                    assembly = type.Assembly;
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs]: Please add brackets for this `if` block.

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs [hunk 3]
```diff
@@ -238,6 +239,182 @@ internal static Assembly LoadGeneratedAssembly(Type type, string defaultNamespac
...
         }
 
+#if XMLSERIALIZERGENERATOR
+        internal static class ThisAssembly
+        {
+            internal const string Version = "1.0.0.0";
+            internal const string InformationalVersion = "1.0.0.0";
+        }
+
+        [ResourceConsumption(ResourceScope.Machine, ResourceScope.Machine)]
```
Comment (mconnew) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: Do these attributes still do something in .Net Core?

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs [hunk 4]
```diff
@@ -238,6 +239,182 @@ internal static Assembly LoadGeneratedAssembly(Type type, string defaultNamespac
...
+            list.Sort();
+            StringBuilder sb = new StringBuilder();
+            for (int i = 0; i < list.Count; i++)
+            {
+                sb.Append(list[i].ToString());
+                sb.Append(",");
+            }
+            return sb.ToString();
+        }
+        internal static FileStream GenerateSerializerFile(XmlMapping[] xmlMappings, Type[] types, string defaultNamespace, Assembly assembly, Hashtable assemblies, string codePath)
```
Comment (mconnew) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: Nit: Blank line between methods

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs [hunk 5]
```diff
@@ -238,6 +239,182 @@ internal static Assembly LoadGeneratedAssembly(Type type, string defaultNamespac
...
+        {
+            Compiler compiler = new Compiler();
+            try
+            {
+                Hashtable scopeTable = new Hashtable();
```
Comment (mconnew) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: Would a Dictionary<,> be a better option here?

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs [hunk 6]
```diff
@@ -238,6 +239,182 @@ internal static Assembly LoadGeneratedAssembly(Type type, string defaultNamespac
...
+                foreach (XmlMapping mapping in xmlMappings)
+                    scopeTable[mapping.Scope] = mapping;
+                TypeScope[] scopes = new TypeScope[scopeTable.Keys.Count];
+                scopeTable.Keys.CopyTo(scopes, 0);
+
+                assemblies.Clear();
+                Hashtable importedTypes = new Hashtable();
```
Comment (mconnew) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: Dictionary?

File: src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs [hunk 7]
```diff
@@ -788,6 +788,46 @@ public static XmlSerializer[] FromMappings(XmlMapping[] mappings, Type type)
...
+                else if (type.Assembly != assembly)
+                {
+                    throw new ArgumentException(SR.Format(SR.XmlPregenOrphanType, type.FullName, assembly.Location), "types");
+                }
+            }
+
+            return TempAssembly.GenerateSerializerFile(mappings, types, null, assembly, new Hashtable(), codePath);
+        }
+#endif
         private static XmlSerializer[] GetSerializersFromCache(XmlMapping[] mappings, Type type)
```
Comment (mconnew) [inline src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs]: Blank line before new method

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs (hunk 3 above)
Comment (huanwu) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: Did a basic test and didn't find anything related. Delete it now.

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs (hunk 5 above)
Comment (huanwu) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: The code is originally from desktop. It is better leave it as it is now.

**Comment by huanwu** [comment]
@dotnet-bot test Windows_NT Debug AllConfigurations Build

**Comment by huanwu** [comment]
@dotnet-bot test Innerloop Ubuntu14.04 Release x64 Build and Test

**Comment by huanwu** [comment]
@dotnet-bot test Innerloop PortableLinux Release x64 Build and Test

File: src/Microsoft.XmlSerializer.Generator/src/Sgen.cs [hunk 8]
```diff
@@ -5,12 +5,305 @@
...
 {
     using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.IO;
+    using System.Linq;
+    using System.Reflection;
+    using System.Threading;
+    using static TempAssembly;
```
Comment (shmao) [inline src/Microsoft.XmlSerializer.Generator/src/Sgen.cs]: Nit: put `using` outside namespace.

File: src/Microsoft.XmlSerializer.Generator/src/Sgen.cs [hunk 9]
```diff
@@ -5,12 +5,305 @@
...
+            sgen.Run(args);
+            return 0;
+        }
+
+        private int Run(string[] args)
+        {
+            string assembly = null;
+            List<string> types = new List<string>();
+            string codePath = null;
+            ArrayList errs = new ArrayList();
```
Comment (shmao) [inline src/Microsoft.XmlSerializer.Generator/src/Sgen.cs]: Nit: var.

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs [hunk 10]
```diff
@@ -238,6 +239,189 @@ internal static Assembly LoadGeneratedAssembly(Type type, string defaultNamespac
...
+
+                FileStream fs;
+                using (fs = File.Create(location))
+                {
+                    string codecontent = compiler.Source.ToString();
+                    Byte[] info = new UTF8Encoding(true).GetBytes(codecontent);
+                    fs.Write(info, 0, info.Length);
+                }
+
+                return fs;
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: `fs` is used in the `using` statements above. The FileStream is already disposed at this point.

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs [hunk 11]
```diff
@@ -238,6 +239,189 @@ internal static Assembly LoadGeneratedAssembly(Type type, string defaultNamespac
...
+                    }
+                }
+
+                for (int i = 0; i < types.Length; i++)
+                {
+                    compiler.AddImport(types[i], importedTypes);
+                }
+
+                compiler.AddImport(typeof(object).Assembly);
+                compiler.AddImport(typeof(XmlSerializer).Assembly);
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: Should we used the full qualified name, `System.Xml.Serialization.XmlSerializer`, here? Or is this `Microsoft.XmlSerializer.Generator.XmlSerializer`?

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs [hunk 12]
```diff
@@ -238,6 +239,189 @@ internal static Assembly LoadGeneratedAssembly(Type type, string defaultNamespac
...
+                writer.WriteLine("[assembly:System.Security.SecurityTransparent()]");
+                writer.WriteLine("[assembly:System.Security.SecurityRules(System.Security.SecurityRuleSet.Level1)]");
+
+                if (assembly != null && types.Length > 0)
+                {
+                    for (int i = 0; i < types.Length; i++)
+                    {
+                        Type type = types[i];
+                        if (type == null)
+                            continue;
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: Nit: empty line.

**Pull request merged**